File: 1040-moving-stones-until-consecutive-ii/1040-moving-stones-until-consecutive-ii.py
class Solution:
    def numMovesStonesII(self, stones: List[int]) -> List[int]:
        stones.sort()
        n = len(stones)
        high = max(stones[-1] - stones[1] - n + 2, stones[-2] - stones[0] - n + 2)
        i, low = 0, n
        for j in range(n):
            while stones[j] - stones[i] >= n:
                i += 1
            if j - i + 1 == n - 1 and stones[j] - stones[i] == n - 2:
                low = min(low, 2)
            else:
                low = min(low, n - (j - i + 1))
        return [low, high]

File: 1040-moving-stones-until-consecutive-ii/1040-moving-stones-until-consecutive-ii.js
/**
 * @param {number[]} stones
 * @return {number[]}
 */
var numMovesStonesII = function(stones) {
  stones.sort((a, b) => a - b);
  const n = stones.length;
  let [j, low] = [0, n];
  let high = Math.max(stones[n - 1] - stones[1], stones[n - 2] - stones[0]) - n + 2;
  
  for (let [i, stone] of stones.entries()) {
    while (stone - stones[j] >= n) j++;
    
    if (i - j === n - 2 && stone - stones[j] === n - 2)
      low = Math.min(low, 2);
    
    else low = Math.min(low, n - (i - j + 1));
  }
  
  return [low, high];
};